### PR TITLE
Attempt to clean up the upload file if it was duplicated/renamed during deployment

### PR DESCRIPTION
**Pull request by benPearce1**
[sc-61082]

This is a workaround which probably requires a better solution.

Related to https://github.com/OctopusDeploy/Issues/issues/7382

### CODE REVIEW

File: source/Calamari.AzureAppService/Behaviors/AzureAppServiceBehaviour.cs [hunk 1]
```diff
@@ -115,35 +116,57 @@ public async Task Execute(RunningDeployment context)
...
+                string token = await Auth.GetAuthTokenAsync(servicePrincipal);
 
-            var webAppClient = new WebSiteManagementClient(new Uri(servicePrincipal.ResourceManagementEndpointBaseUri),
-                    new TokenCredentials(token))
-                {SubscriptionId = servicePrincipal.SubscriptionNumber};
+                var webAppClient = new WebSiteManagementClient(new Uri(servicePrincipal.ResourceManagementEndpointBaseUri),
+                                                               new TokenCredentials(token))
+                {
+                    SubscriptionId = servicePrincipal.SubscriptionNumber
+                };
```
Comment (benPearce1) [inline source/Calamari.AzureAppService/Behaviors/AzureAppServiceBehaviour.cs:154]: formatting

File: source/Calamari.AzureAppService/Behaviors/AzureAppServiceBehaviour.cs [hunk 2]
```diff
@@ -115,35 +116,57 @@ public async Task Execute(RunningDeployment context)
...
-            await UploadZipAsync(publishingProfile, httpClient, uploadPath, targetSite.ScmSiteAndSlot);
+                await UploadZipAsync(publishingProfile, httpClient, uploadPath, targetSite.ScmSiteAndSlot);
+            }
+            finally
+            {
+                if (uploadFileNeedsCleaning)
+                {
+                    CleanupUploadFile(uploadPath);
+                }
+            }
```
Comment (benPearce1) [inline source/Calamari.AzureAppService/Behaviors/AzureAppServiceBehaviour.cs:169]: always run even if there is an exception

File: source/Calamari.AzureAppService/Behaviors/AzureAppServiceBehaviour.cs [hunk 3]
```diff
@@ -194,5 +217,13 @@ private async Task<IDeploymentSlot> FindOrCreateSlot(IWebApp client, TargetSite
 
             Log.Verbose("Finished deploying");
         }
+        
+        void CleanupUploadFile(string? uploadPath)
+        {
+            if (File.Exists(uploadPath))
```
Comment (benPearce1) [inline source/Calamari.AzureAppService/Behaviors/AzureAppServiceBehaviour.cs:223]: `File.Exists` has built-in nullability checks

File: source/Calamari.AzureAppService/Behaviors/AzureAppServiceBehaviour.cs [hunk 4]
```diff
@@ -115,35 +116,57 @@ public async Task Execute(RunningDeployment context)
...
+                if (substitutionFeatures.Any(featureName => context.Variables.IsFeatureEnabled(featureName)))
+                {
+                    uploadPath = (await Archive.PackageArchive(context.StagingDirectory, context.CurrentDirectory)).FullName;
+                    uploadFileNeedsCleaning = false;
+                }
+                else
+                {
+                    var uploadFile = await Archive.ConvertToAzureSupportedFile(packageFileInfo); 
+                    uploadPath = uploadFile.FullName;
+                    uploadFileNeedsCleaning = uploadFile.Extension != packageFileInfo.Extension; // did the file get changed
```
Comment (benPearce1) [inline source/Calamari.AzureAppService/Behaviors/AzureAppServiceBehaviour.cs:131]: There are three implementations of the package providers, only the nuget one makes any changes to the file for uploading.

Comment (IsaacCalligeros95) [reply source/Calamari.AzureAppService/Behaviors/AzureAppServiceBehaviour.cs:131]: There's also the LegazyAzureAppServiceBehaviour that follows this same code path. 
Can the file deletion be pushed into the NugetPackageProvider to handle both?

Comment (IsaacCalligeros95) [reply source/Calamari.AzureAppService/Behaviors/AzureAppServiceBehaviour.cs:131]: Scrap that this is 2023.3. LGTM

**Review by IsaacCalligeros95 — APPROVED** [review]
LGTM

File: source/Calamari.AzureAppService/Behaviors/AzureAppServiceBehaviour.cs [hunk 5]
```diff
@@ -5,6 +5,7 @@
 using System.Linq;
 using System.Net.Http;
 using System.Net.Http.Headers;
+using System.Runtime.CompilerServices;
```
Comment (IsaacCalligeros95) [inline source/Calamari.AzureAppService/Behaviors/AzureAppServiceBehaviour.cs:8]: Is this required?

**Pull request merged**